cmd/zdb/zdb.c

```diff
@@ -9112,23 +9137,63 @@ main(int argc, char **argv)
 	if (argc < 2 && dump_opt['R'])
 		usage();
 
+	target = argv[0];
+
+	/*
+	 * Automate cachefile
+	 */
+	if (!spa_config_path_env && !config_path_console && target) {
+		zpool_handle_t *zhp = NULL;
+		char prop[ZFS_MAXPROPLEN];
+		zprop_source_t srctype = ZPROP_SRC_LOCAL;
+		char *pname = strdup(target);
+
+		if (strpbrk(pname, "/@") != NULL)
+			*strpbrk(pname, "/@") = '\0';
+
+		if ((g_zfs = libzfs_init()) && (zhp = zpool_open_canfail(g_zfs,
+		    pname)) && !zpool_get_prop(zhp, ZPOOL_PROP_CACHEFILE, prop,
+		    sizeof (prop), &srctype, B_FALSE) && (prop[0] != '\0')) {
+			if (prop[0] == '/') {
+				if (access(prop, F_OK) == 0)
+					spa_config_path = prop;
+				else
+					force_import = B_TRUE;
+			} else if ((strcmp(prop, "-") == 0 &&
+			    access(ZPOOL_CACHE, F_OK) != 0) ||
+			    strcmp(prop, "none") == 0) {
+				force_import = B_TRUE;
+			}
+		}
+		if (zhp != NULL)
+			zpool_close(zhp);
+		free(pname);
+	}
+
+	kernel_init(SPA_MODE_READ);
```

Comment:
Moving the `kernel_init()` call here seems like the most likely cause the ASSERT we're now hitting reliably in `alloc_class/alloc_class_012_pos`.

Reply:
@behlendorf - Today I got a chance to dig deeper, and it appears that it only fails when ZFS is compiled from RPM packaging, consistently triggering this ASSERT: https://github.com/openzfs/zfs/blob/master/module/zcommon/zfs_fletcher.c#L429. Interestingly, even if I revert the entire patch and just keep the inclusion of `libzfs.la` in the `cmd/zdb/Makefile`, the issue still reproduces. I don't have much experience with RPM packaging, but I'll try to figure it out. Any suggestions would be appreciated.